lib/Dialect/RTL/Ops.cpp

```diff
@@ -697,6 +744,19 @@ void MulOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
 
       APInt value, value2;
 
+      // mul(x, c) -> mul( shl(x, log2(c)) ), where c is a power of two.
+      if (size == 2 && matchPattern(inputs.back(), m_RConstant(value)) &&
+          value.isPowerOf2()) {
+        auto shift =
+            rewriter.create<ConstantOp>(op.getLoc(), value.exactLogBase2(),
+                                        op.getType().cast<IntegerType>());
+        auto shlOp = rewriter.create<rtl::ShlOp>(op.getLoc(), inputs[0], shift);
+
+        std::array<Value, 1> newOperands = {shlOp};
+        rewriter.replaceOpWithNewOp<MulOp>(op, op.getType(), newOperands);
```

Comment:
In this case, I'd recommend seeing if you can pass this as `ArrayRef(shlOp)` instead of the intermediate std::array.